### PR TITLE
PreserveCompilationContext should take into account runtime store assemblies

### PR DESCRIPTION
**Pull request by eerhardt**
When PreserveCompilationContext is set to `true`, runtime assemblies that are already in the runtime store shouldn't be copied to the `refs` folder.  Instead, they should be resolved from the runtime store at runtime.

Fix https://github.com/dotnet/sdk/issues/1216

End-to-end scenario is blocked by https://github.com/dotnet/core-setup/pull/2498

### CODE REVIEW

**Comment by eerhardt** [comment]
test Windows_NT_FullFramework Release

File: src/Tasks/Microsoft.NET.Build.Tasks/FindItemsFromPackages.cs [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+        [Required]
+        public ITaskItem[] Items { get; set; }
+
+        [Required]
+        public ITaskItem[] Packages { get; set; }
+
+        [Output]
+        public ITaskItem[] ItemsFromPackages
+        {
+            get { return _itemsFromPackages.ToArray(); }
```
Comment (nguerrera) [inline src/Tasks/Microsoft.NET.Build.Tasks/FindItemsFromPackages.cs]: I guess msbuild only calls this once per task invocation, but calling ToArray() in a getter is usually a bad idea in the general case. Looks like field access, but is really O(N). 

Perhaps this is better:

``` C#
public ITaskItem[] ItemsFromPackages { get; private set; }

protected override ExecuteCore()
{
    var list = new List<ITaskItem>();
    // ... add to list ...
   ItemsFromPackages = list.ToArray();
}
```

File: test/Microsoft.NET.Publish.Tests/GivenThatWeWantToPreserveCompilationContext.cs [hunk 2]
```diff
@@ -123,11 +127,65 @@ public void It_publishes_the_project_with_a_refs_folder_and_correct_deps_file()
...
+                "<StoreArtifacts>",
+                @"  <Package Id=""Newtonsoft.Json"" Version=""9.0.1"" />",
+                @"  <Package Id=""System.Data.Common"" Version=""4.3.0"" />",
+                @"  <Package Id=""System.Data.SqlClient"" Version=""4.3.0"" />",
+                "</StoreArtifacts>",
+            });
+
+            var appProjectDirectory = Path.Combine(testAsset.TestRoot, "TestApp");
+
+            var targetFramework = "netcoreapp1.1";
```
Comment (dsplaisted) [inline test/Microsoft.NET.Publish.Tests/GivenThatWeWantToPreserveCompilationContext.cs]: Is the runtime store feature even supposed to be enabled for .NET Core 1.x?

**Review by dsplaisted — APPROVED** [review]

File: test/Microsoft.NET.Publish.Tests/GivenThatWeWantToPreserveCompilationContext.cs [hunk 3]
```diff
@@ -286,9 +346,5 @@ public void It_publishes_the_project_with_a_refs_folder_and_correct_deps_file()
 System.Xml.ReaderWriter.dll
 System.Xml.XDocument.dll
 TestLibrary.dll".Split(new char[] { '\r', '\n' }, StringSplitOptions.RemoveEmptyEntries);
-
-        public GivenThatWeWantToPreserveCompilationContext(ITestOutputHelper log) : base(log)
-        {
-        }
```
Comment (nguerrera) [inline test/Microsoft.NET.Publish.Tests/GivenThatWeWantToPreserveCompilationContext.cs]: Thanks for moving this. I generated these with (Ctrl + '.'). I don't know why this one didn't go at the top like the others.

**Review by nguerrera — APPROVED** [review]

File: src/Tasks/Microsoft.NET.Build.Tasks/FindItemsFromPackages.cs (hunk 1 above)
Comment (eerhardt) [inline src/Tasks/Microsoft.NET.Build.Tasks/FindItemsFromPackages.cs]: I'll fix this in the new files I'm adding.  Note that this is a pattern used in about half the places this repo.

Comment (nguerrera) [inline src/Tasks/Microsoft.NET.Build.Tasks/FindItemsFromPackages.cs]: OK

File: test/Microsoft.NET.Publish.Tests/GivenThatWeWantToPreserveCompilationContext.cs (hunk 2 above)
Comment (eerhardt) [inline test/Microsoft.NET.Publish.Tests/GivenThatWeWantToPreserveCompilationContext.cs]: You're right.  Moved to netcoreapp2.0.

**Comment by eerhardt** [comment]
**Customer scenario**

When ASP.NET apps don't have Razor pre-compilation turned on, we copy the "compile-time" reference assemblies to a "refs" folder in the publish folder.  Normal assemblies that are used for both compile-time and runtime are not normally copied to the "refs" folder, because they can be resolved from the normal publish folder.
With the introduction of the "runtime store", all of the ASP.NET runtime assemblies are no longer in the normal publish folder.  Instead they are contained in the "runtime store".  But, we never updated the code that copies the files to the "refs" folder to respect the "runtime store".  So now all the ASP.NET runtime assemblies are also copied to the "refs" folder, making the published app bigger than it needs to be.

**Bugs this fixes:** 

#1216 

**Workarounds, if any**

None really.  Users could try to implement this fix in their MSBuild projects themselves, but this affects any ASP.NET app that doesn't use pre-compilation of Razor views.

**Risk**

Medium risk for this whole scenario.  We've had quite a few bugs in this area (resolving compile-time assemblies at runtime).  For this specific fix - Medium-Low. I had to refactor how we determine which packages are in the runtime store, but I think this refactoring makes for cleaner code going forward.

**Performance impact**

This makes published apps smaller, since they won't need to copy all of the ASP.NET runtime assemblies.

**Is this a regression from a previous update?**
No

**Root cause analysis:**
We missed this cross-feature impact when accounting for the runtime store feature.  This was caused by 2 features interacting together: PreserveCompilationContext + Runtime Store.

**How was the bug found?**
Found by @DamianEdwards on the ASP.NET team.  I assume it was from him dogfooding the product.


@MattGertz for approval

Note - If the release/2.0.0 branch opens up before merging this, I will retarget to that branch.  If not, I'll merge to master and port my change.

**Comment by eerhardt** [comment]
test Windows_NT_FullFramework Debug
test Windows_NT_FullFramework Release

**Comment by eerhardt** [comment]
test Windows_NT_FullFramework Debug

**Pull request merged**